### PR TITLE
Allow more advanced test case support semantic

### PR DESCRIPTION
**Pull request by lidizheng**
After this PR, the framework can skip a test case for a specific language with a specific version range, or something more fancy.

Test runs:

- cpp: https://fusion2.corp.google.com/invocations/2132293d-5c36-444c-83a6-be522f06d16d/targets
- python: https://fusion2.corp.google.com/invocations/d57ffd03-9b51-495d-ac7d-b1ef67626f1b/targets

### CODE REVIEW

File: tools/run_tests/xds_k8s_test_driver/framework/xds_flags.py [hunk 1]
```diff
@@ -109,6 +109,12 @@
         "XdsStats, XdsUpdateClientConfigure, and ProtoReflection (optional).\n"
         "Doesn't have to be within --firewall_allowed_ports."))
 
+# Testing metadata
+TESTING_VERSION = flags.DEFINE_string(
+    "testing_version",
+    default="master",
+    help="The testing gRPC version branch name.")
```
Comment (menghanl) [inline tools/run_tests/xds_k8s_test_driver/framework/xds_flags.py:116]: Add example values to the help string.

File: tools/run_tests/xds_k8s_test_driver/framework/xds_url_map_testcase.py [hunk 2]
```diff
@@ -233,22 +251,16 @@ class XdsUrlMapTestCase(absltest.TestCase, metaclass=_MetaXdsUrlMapTestCase):
...
-        return _CLIENT_LANGUAGES
-
-    @staticmethod
-    def supported_servers() -> Tuple[str]:
-        """Declare supported languages of servers.
+        The config includes client_lang, server_lang, version.
 
         Returns:
-          A tuple of strings contains the supported languages for this test.
+          A tuple of strings contains the supported versions for this test or
```
Comment (menghanl) [inline tools/run_tests/xds_k8s_test_driver/framework/xds_url_map_testcase.py:260]: Is this doc right?

File: tools/run_tests/xds_k8s_test_driver/tests/url_map/retry_test.py [hunk 3]
```diff
@@ -66,8 +67,10 @@ def _build_retry_route_rule(retryConditions, num_retries):
 class TestRetryUpTo3AttemptsAndFail(xds_url_map_testcase.XdsUrlMapTestCase):
 
     @staticmethod
-    def supported_clients() -> Tuple[str]:
-        return 'cpp', 'java', 'go'
+    def is_supported(config: TestConfig) -> bool:
+        if config.client_lang in ['cpp', 'java', 'go']:
+            return config.version_ge('v1.40.x')
```
Comment (menghanl) [inline tools/run_tests/xds_k8s_test_driver/tests/url_map/retry_test.py:72]: Retry should be the same as affinity. go version >= 1.41.x

**Review by menghanl — CHANGES_REQUESTED** [review]

File: tools/run_tests/xds_k8s_test_driver/framework/xds_flags.py (hunk 1 above)
Comment (lidizheng) [reply tools/run_tests/xds_k8s_test_driver/framework/xds_flags.py]: Done. Good idea.

File: tools/run_tests/xds_k8s_test_driver/framework/xds_url_map_testcase.py (hunk 2 above)
Comment (lidizheng) [reply tools/run_tests/xds_k8s_test_driver/framework/xds_url_map_testcase.py]: Awkward, updated.

File: tools/run_tests/xds_k8s_test_driver/tests/url_map/retry_test.py (hunk 3 above)
Comment (lidizheng) [reply tools/run_tests/xds_k8s_test_driver/tests/url_map/retry_test.py:72]: Done.

File: tools/run_tests/xds_k8s_test_driver/tests/url_map/retry_test.py [hunk 4]
```diff
@@ -105,8 +110,10 @@ def rpc_distribution_validate(self, test_client: XdsTestClient):
 class TestRetryUpTo4AttemptsAndSucceed(xds_url_map_testcase.XdsUrlMapTestCase):
 
     @staticmethod
-    def supported_clients() -> Tuple[str]:
-        return 'cpp', 'java', 'go'
+    def is_supported(config: TestConfig) -> bool:
+        if config.client_lang in ['cpp', 'java', 'go']:
+            return config.version_ge('v1.40.x')
```
Comment (menghanl) [inline tools/run_tests/xds_k8s_test_driver/tests/url_map/retry_test.py:115]: And this

**Review by menghanl — APPROVED** [review]

**Pull request merged**